src/main/scala/uk/gov/ons/bi/dataload/ubrn/LinksPreprocessor.scala

```diff
@@ -0,0 +1,91 @@
+package uk.gov.ons.bi.dataload.ubrn
+
+import com.google.inject.Singleton
+import org.apache.spark.SparkContext
+import org.apache.spark.sql.{DataFrame, SQLContext}
+import org.apache.spark.sql.functions._
+import uk.gov.ons.bi.dataload.reader.LinkJsonReader
+import uk.gov.ons.bi.dataload.utils.AppConfig
+
+import scala.util.{Success, Try}
+
+/**
+  * Created by websc on 03/03/2017.
+  */
+
+@Singleton
+class LinksPreprocessor(sc: SparkContext) {
+
+  val defaultBaseUbrn = 1L
+
+  def getMaxUbrn(df: DataFrame, ubrnColName: String = "UBRN"): Option[Long] = {
+    // This will scan the DF column to extract the max value, assumes values are numeric.
+    // Defaults to zero.
+    Try {
+      val row = df.agg(max(df(ubrnColName))).collect.head
+      row.getLong(0)
+    }
+    match {
+      case Success(n: Long) => Some(n)
+      case _ => Some(defaultBaseUbrn)
+    }
+  }
+
+  def applyNewUbrn(df: DataFrame, baseUbrn: Option[Long] = None) = {
+    // First drop any rogue UBRN column (if any) from the input DF
+    val noUbrn = df.drop("UBRN")
+
+    // Set the base UBRN for adding to the monotonic sequential value
+    val base = baseUbrn.getOrElse(defaultBaseUbrn)
+
+    // Repartition to one partition so sequence is a fairly continuous range.
+    // This will force data to be shuffled, which is inefficient.
+    val numPartitions = df.rdd.getNumPartitions
+
+    val df1partition = df.repartition(1)
```

Comment:
Yes, this could be a problem, although Spark can spill data to disk as this should be a "transformation", not an "action" (no need to collect data on driver).  This is the only way to keep the numbers inside a single range using the Spark Monotonically Increasing ID function. If we partition the data, then we will get multiple ranges of IDs, which management seem to dislike.